### ISSUE TITLE
DISABLED test_ref_small_input_prod_cuda_int16 (__main__.TestReductionsCUDA)

### ISSUE THREAD

**Issue by suo** [pytorch/pytorch#77305]
Platforms: win

This test was disabled because it is failing on master ([recent examples](http://torch-ci.com/failure/test_ref_small_input_prod_cuda_int8%2C%20TestReductionsCUDA)).

cc @peterjc123 @mszhanyi @skyline75489 @nbcsm @mruberry

**Comment by suo** [comment]
Background: these errors started popping up after https://github.com/pytorch/pytorch/commit/eaee4ff9ed40376043ab6d8eefc0874b6ea38c28.

@mruberry: it's a bit surprising to me that setting `PYTORCH_TESTING_DEVICE_ONLY_FOR` would cause more tests to be run in some cases, instead of strictly fewer. Do you know what's going on there?

**Comment by mruberry** [comment]
That's causing more tests to run? That seems wrong. The code here should just instantiate fewer test bases: 

https://github.com/pytorch/pytorch/blob/2881e0ea174162179cfc7a7a6d75de952d60fce8/torch/testing/_internal/common_device_type.py#L601

**Comment by suo** [comment]
Yes, it's very strange. Taking `test_ref_small_input_prod_cuda_int8` as an example:
- Commit right before my change (https://github.com/pytorch/pytorch/commit/4f4ebc549104f79eda719a7712a49e83535e0902): https://fburl.com/scuba/pytorch_test_times/mwq9spdn. As you can see, the test does not run (empty data).
- Then it starts running at https://github.com/pytorch/pytorch/commit/eaee4ff9ed40376043ab6d8eefc0874b6ea38c28: https://fburl.com/scuba/pytorch_test_times/zpmwbbgs.

**Comment by ngimel** [comment]
There's a weirdness with test instantiation. Locally, if I don't set PYTORCH_TESTING_DEVICE_ONLY_FOR, cuda tests are not instantiated at all for that test, with a variable they are instantiated:
``` 
(pytorch) [ngimel@devgpu006.prn1 ~/local/pytorch/test] python test_reductions.py -v -k test_ref_small_input_prod_cuda

----------------------------------------------------------------------
Ran 0 tests in 0.000s

OK
(pytorch) [ngimel@devgpu006.prn1 ~/local/pytorch/test] PYTORCH_TESTING_DEVICE_ONLY_FOR="cuda" python test_reductions.py -v -k test_ref_small_input_prod_cuda
test_ref_small_input_prod_cuda_bool (__main__.TestReductionsCUDA)
Compares op against reference for small input tensors ... ok
test_ref_small_input_prod_cuda_complex128 (__main__.TestReductionsCUDA)
Compares op against reference for small input tensors ... ok
test_ref_small_input_prod_cuda_complex64 (__main__.TestReductionsCUDA)
Compares op against reference for small input tensors ... skipped 'Skipped!'
test_ref_small_input_prod_cuda_float16 (__main__.TestReductionsCUDA)
Compares op against reference for small input tensors ... skipped 'Skipped!'
test_ref_small_input_prod_cuda_float32 (__main__.TestReductionsCUDA)
Compares op against reference for small input tensors ... ok
test_ref_small_input_prod_cuda_float64 (__main__.TestReductionsCUDA)
Compares op against reference for small input tensors ... ok
test_ref_small_input_prod_cuda_int16 (__main__.TestReductionsCUDA)
Compares op against reference for small input tensors ... ok
test_ref_small_input_prod_cuda_int32 (__main__.TestReductionsCUDA)
Compares op against reference for small input tensors ... ok
test_ref_small_input_prod_cuda_int64 (__main__.TestReductionsCUDA)
Compares op against reference for small input tensors ... ok
test_ref_small_input_prod_cuda_int8 (__main__.TestReductionsCUDA)
Compares op against reference for small input tensors ... ok
test_ref_small_input_prod_cuda_uint8 (__main__.TestReductionsCUDA)
Compares op against reference for small input tensors ... ok

----------------------------------------------------------------------
Ran 11 tests in 10.472s

OK (skipped=2)
```

**Comment by ngimel** [comment]
so if ops are filtered, like `@ops(filter(lambda op: op.ref is not None, reduction_ops))`, like any iterator this filter works once, for the first device, and all subsequent devices instantiate 0 tests.

**Comment by ngimel** [comment]
Test instantiation fixed in #77330
